src/mlpack/methods/lsh_model/lshmodel.hpp

```diff
@@ -0,0 +1,373 @@
+
+   //! Return the reference set.
+   const arma::mat ReferenceSet(void) const {return *referenceSet; };
+
+   //! Serialize the LSHModel object.
+   template<typename Archive>
+   void Serialize(Archive& ar);
+
+ private:
+
+  /**
+   * Returns the score of a perturbation vector generated by perturbation set A.
+   * The score of a pertubation set (vector) is the sum of scores of the
+   * participating actions.
+   * @param A perturbation set to compute the score of.
+   * @param scores vector containing score of each perturbation.
+  */
+  double PerturbationScore(const std::vector<bool>& A,
+                                  const arma::vec& scores) const;
+  /**
+   * Inline function used by GetAdditionalProbingBins. The vector shift operation
+   * replaces the largest element of a vector A with (largest element) + 1.
+   * Returns true if resulting vector is valid, otherwise false.
+   * @param A perturbation set to shift.
+  */
+  bool PerturbationShift(std::vector<bool>& A) const;
+
+  /**
+   * Inline function used by GetAdditionalProbingBins. The vector expansion
+   * operation adds the element [1 + (largest_element)] to a vector A, where
+   * largest_element is the largest element of A. Returns true if resulting vector
+   * is valid, otherwise false.
+   * @param A perturbation set to expand.
+  */
+  bool PerturbationExpand(std::vector<bool>& A) const;
+
+  /**
+   * Return true if perturbation set A is valid. A perturbation set is invalid if
+   * it contains two (or more) actions for the same dimension or dimensions that
+   * are larger than the queryCode's dimensions.
+   * @param A perturbation set to validate.
+   * @param numProj The number of projections for the sequence under validation.
+  */
+  bool PerturbationValid(const std::vector<bool>& A, size_t numProj) const;
+   /**
+    * Function that creates a template perturbation sequence given a value for
+    * an M and a W. The template perturbation sequence is based on the
+    * statistical properties of multi-probe LSH and uses those, instead of
+    * specific points, to generate scores.
+    * See mlpack/methods/lsh/lsh_search_impl.hpp for more details about how
+    * perturbation sequences are generated from specific points.
+    *
+    * @param numProj The number of projections for the LSH scheme for which we
+    *     want to compute the template perturbation sequence.
+    * @param hashWidth The hash width for the LSH scheme.
+    * @param numProbes The number of probes to generate.
+    */
+   void GenerateTemplateSequence(size_t numProj, 
+                                 double hashWidth, 
+                                 size_t numProbes);
+
+   /** Matrix that stores, in each column, the "direction" of the perturbation:
+    * 0 means no perturbation on that dimension, -1 means reduce dimension value
+    * by 1, and +1 means increase dimension value by 1.
+    */
+   
+   arma::Mat<short int> templateSequence;
```

Comment:
Consider `std::vector<std::vector<bool>>` here, it might be more space-efficient and possibly quicker.  Also a possibility is `boost::dynamic_bitset` instead of `std::vector<bool>`, but I am not sure that gets you anything over `std::vector<bool>` here.

Yet another thought is that generally these sequences will be pretty sparse, so it might be more reasonable to encode each perturbation sequence as a 2-dimensional matrix: one dimension represents the nonzero coordinate and the other dimension represents whether the value is positive or negative.  For that matter you could be tricky and encode both the coordinate and the direction as a single number (i.e. direction*coordinate), but then you could only have number of coordinates equal to half the possible number of coordinates of that data type (but I don't think that is a problem).